kafka-rest/src/main/java/io/confluent/kafkarest/controllers/AbstractConfigManager.java

```diff
@@ -0,0 +1,157 @@
+import io.confluent.kafkarest.entities.AbstractConfig;
+import io.confluent.kafkarest.entities.ConfigSource;
+import io.confluent.kafkarest.entities.ConfigSynonym;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.admin.AlterConfigOp;
+import org.apache.kafka.clients.admin.ConfigEntry;
+import org.apache.kafka.clients.admin.DescribeConfigsOptions;
+import org.apache.kafka.common.config.ConfigResource;
+
+/**
+ * An abstract base class for managers of subtypes of {@link AbstractConfig}.
+ */
+abstract class AbstractConfigManager<
+    T extends AbstractConfig, B extends AbstractConfig.Builder<T, B>> {
+
+  private final Admin adminClient;
+  private final ClusterManager clusterManager;
+
+  AbstractConfigManager(Admin adminClient, ClusterManager clusterManager) {
+    this.adminClient = requireNonNull(adminClient);
+    this.clusterManager = requireNonNull(clusterManager);
+  }
+
+  final CompletableFuture<List<T>> listConfigs(
+      String clusterId, ConfigResource resourceId, B prototype) {
+    return clusterManager.getCluster(clusterId)
+        .thenApply(cluster -> checkEntityExists(cluster, "Cluster %s cannot be found.", clusterId))
+        .thenCompose(
+            cluster ->
+                KafkaFutures.toCompletableFuture(
+                    adminClient.describeConfigs(
+                        singletonList(resourceId),
+                        new DescribeConfigsOptions().includeSynonyms(true))
+                        .values()
+                        .get(resourceId)))
+        .thenApply(
+            response ->
+                response.entries().stream()
+                    .map(
+                        entry ->
+                            prototype.setName(entry.name())
+                                .setValue(entry.value())
+                                .setDefault(entry.isDefault())
+                                .setReadOnly(entry.isReadOnly())
+                                .setSensitive(entry.isSensitive())
+                                .setSource(ConfigSource.fromAdminConfigSource(entry.source()))
+                                .setSynonyms(
+                                    entry.synonyms().stream()
+                                        .map(ConfigSynonym::fromAdminConfigSynonym)
+                                        .collect(Collectors.toList()))
+                                .build())
+                    .collect(Collectors.toList()));
+  }
+
+  final CompletableFuture<Optional<T>> getConfig(
+      String clusterId, ConfigResource resourceId, B prototype, String name) {
+    return listConfigs(clusterId, resourceId, prototype)
+        .thenApply(configs -> findEntityByKey(configs, AbstractConfig::getName, name));
+  }
+
+  final CompletableFuture<Void> safeUpdateConfig(
```

Comment:
Shall we add comment to describe the safe and unsafe way to update/reset configs ?

Reply:
Done.